client/starwhale/core/dataset/store.py

```diff
@@ -438,19 +453,151 @@ def sign_uri(self, uri: str) -> str:
+
+    def read(self, size: int) -> bytes:
+        ret = self._read(size)
+        if isinstance(ret, memoryview):
+            return ret.tobytes()
+        elif isinstance(ret, bytes):
+            return ret
+        else:
+            raise FormatError(f"{ret}:{type(ret)} is not bytes or memoryview type")
+
+    def _read(self, size: int) -> t.Union[memoryview, bytes]:
+        if size == 0:
+            return b""
+        elif size < 0:
+            return self._read_buffer.read() + self._read_exhausted_raw()
+        else:
+            while len(self._read_buffer) < size:
+                fill_size = self._fill_new_bytes()
+                if fill_size == 0:
+                    return self._read_buffer.read()
+
+            return self._read_buffer.read(size)
+
+    @abstractmethod
+    def _read_exhausted_raw(self) -> bytes:
+        raise NotImplementedError
+
+    @abstractmethod
+    def _fill_new_bytes(self) -> int:
+        raise NotImplementedError
+
+    def close(self) -> None:
+        self._read_buffer.close()
+
+
+class HttpBufferedFileLike(BaseBufferedFileLike):
+    def __init__(
+        self,
+        url: str,
+        headers: t.Optional[t.Dict[str, str]] = None,
+        timeout: int = 90,
+        buffer_size: int = 256 * 1024,
+    ) -> None:
+        super().__init__(buffer_size)
+        if headers is None:
+            self.headers = {"Accept-Encoding": "identity"}
+        else:
+            self.headers = headers
+
+        self.buffer_size = buffer_size
+        self.resp = requests.get(
+            url,
+            stream=True,
+            headers=self.headers,
+            timeout=timeout,
+        )
+
+        if not self.resp.ok:
+            self.resp.raise_for_status()
+
+        self._read_iter = self.resp.iter_content(buffer_size)
+
+    def _read_exhausted_raw(self) -> bytes:
+        return self.resp.raw.read()  # type: ignore
+
+    def _fill_new_bytes(self) -> int:
+        return self._read_buffer.write_from_iter(self._read_iter)
+
+    def close(self) -> None:
+        self.resp.close()
+        return super().close()
+
+
+class BytesBuffer:
+    def __init__(self, size: int = io.DEFAULT_BUFFER_SIZE) -> None:
+        self._size = size
+        self._pos = 0
+        self._buffer = memoryview(bytearray(0))
+
+    def __str__(self) -> str:
+        return f"buffer size: {self._size}, pos: {self._pos}, length: {len(self)}"
+
+    __repr__ = __str__
+
+    def close(self) -> None:
+        self._buffer.release()
+
+    def __len__(self) -> int:
+        return len(self._buffer) - self._pos
+
+    def read(self, size: int = -1) -> bytes:
+        if size < 0 or size > len(self):
+            size = len(self)
+
+        ret = self._buffer[self._pos : self._pos + size]
+        self._pos += size
+        return ret.tobytes()
+
+    def write(self, new_bytes: bytes) -> int:
+        _release_buffer = self._buffer
+        self._buffer = memoryview(self._buffer[self._pos :].tobytes() + new_bytes)
```

Comment:
`self._buffer[self._pos :].tobytes() + new_bytes` will allocate new memory and create another copy of the data, is this expected?

Reply:
updated,  use `bytearray` directly. However, sometimes write function needs allocate new memory.